src/RecursiveIteratorTrait.php

```diff
@@ -32,49 +33,98 @@
     protected $pathSegments;
 
     /**
-     * Resets the values.
+     * Computes a reset state.
      *
      * @since [*next-version*]
+     *
+     * @return IterationInterface The iteration that represents the new state.
      */
-    protected function _construct()
+    protected function _reset()
     {
         $this->_resetParents();
+        $this->_pushParent($this->_getInitialParentIterable());
 
-        parent::_construct();
+        return $this->_loop();
     }
 
     /**
-     * Retrieves the current path segments.
+     * Advances the iterator and computes the new state.
      *
      * @since [*next-version*]
      *
-     * @return array
+     * @return IterationInterface The iteration that represents the new state.
      */
-    protected function _getPathSegments()
+    protected function _loop()
     {
-        return $this->pathSegments;
+        // Ensure that there are items on the stack
+        if (!$this->_hasParents()) {
+            return $this->_createIteration(null, null);
+        }
+
+        // Get current top item on the stack and its current iteration entry
+        $parent = &$this->_getCurrentIterable();
+        $current = $this->_createCurrentIteration($parent);
+
+        // Reached end of current iterable
+        if ($current->getKey() === null) {
+            return $this->_backtrackLoop();
+        }
+
+        // Element is a leaf
+        if (!$this->_isElementHasChildren($current->getValue())) {
+            next($parent);
+
+            return $current;
+        }
+
+        // Element is not a leaf; push to stack
+        $children = $current->getValue();
+        $this->_pushParent($children);
+
+        if ($this->_isMode(R::MODE_SELF_FIRST)) {
```

Comment:
Didn't find an implementation of this, which would probably have a contract for `_getMode()` (already exists).

Reply:
You're right. Should I remove this and just call `_getMode()` and compare?

Reply:
No, I meant that perhaps there should be an `IsModeCapableTrait`.

Reply:
Oh. Don't you think it's simpler to just compare though? I mean, it's an extra trait to implement, test and `use` in classes, all to just compare.

Reply:
Then again, determining mode would be abstracted, and not tied to comparing a result from another method (`_getMode()`).

Reply:
Thing is, for most cases you don't need to _know_ the mode; only to _compare_ the mode. Therefore, there may be no "saved" mode, but comparison is done using an algorithm. This is much more flexible.